File: lib_openweather/src/main/java/guepardoapps/library/openweather/controller/ReceiverController.java
package guepardoapps.library.openweather.controller;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.IntentFilter;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import guepardoapps.library.openweather.utils.Logger;

@SuppressWarnings({"WeakerAccess"})
public class ReceiverController implements IReceiverController {
    private static String Tag = ReceiverController.class.getSimpleName();

    private Context _context;
    private List<BroadcastReceiver> _registeredReceiver;

    public ReceiverController(@NonNull Context context) {
        _context = context;
        _registeredReceiver = new ArrayList<>();
    }

    @Override
    public void RegisterReceiver(@NonNull BroadcastReceiver receiver, @NonNull String[] actions) {
        IntentFilter downloadStateFilter = new IntentFilter();
        for (String action : actions) {
            downloadStateFilter.addAction(action);
        }

        UnregisterReceiver(receiver);

        _context.registerReceiver(receiver, downloadStateFilter);
        _registeredReceiver.add(receiver);
    }

    @Override
    public void UnregisterReceiver(@NonNull BroadcastReceiver receiver) {
        for (int index = 0; index < _registeredReceiver.size(); index++) {
            if (_registeredReceiver.get(index) == receiver) {
                try {
                    _context.unregisterReceiver(receiver);
                    _registeredReceiver.remove(index);
                } catch (Exception exception) {
                    Logger.getInstance().Error(Tag, exception.toString());
                }
                break;
            }
        }
    }

    @Override
    public void Dispose() {
        for (int index = 0; index < _registeredReceiver.size(); index++) {
            try {
                _context.unregisterReceiver(_registeredReceiver.get(index));
                _registeredReceiver.remove(index);
            } catch (Exception exception) {
                Logger.getInstance().Error(Tag, exception.toString());
            }
        }
    }
}

File: app/src/main/java/guepardoapps/lib/openweather/controller/ReceiverController.kt
package guepardoapps.lib.openweather.controller

import android.content.BroadcastReceiver
import android.content.Context
import android.content.IntentFilter
import guepardoapps.lib.openweather.utils.Logger

class ReceiverController(private val context: Context) : IReceiverController {
    private val tag: String = ReceiverController::class.java.canonicalName

    private val registeredReceiver: MutableList<BroadcastReceiver> = mutableListOf()

    override fun registerReceiver(broadcastReceiver: BroadcastReceiver, actions: Array<String>) {
        val intentFilter = IntentFilter()

        for (action in actions) {
            intentFilter.addAction(action)
        }

        unregisterReceiver(broadcastReceiver)

        context.registerReceiver(broadcastReceiver, intentFilter)
        registeredReceiver.add(broadcastReceiver)
    }

    override fun unregisterReceiver(broadcastReceiver: BroadcastReceiver) {
        for (receiver in registeredReceiver) {
            if (receiver === broadcastReceiver) {
                try {
                    context.unregisterReceiver(receiver)
                    registeredReceiver.remove(receiver)
                } catch (exception: Exception) {
                    Logger.instance.error(tag, exception)
                }
                break
            }
        }
    }

    override fun dispose() {
        for (receiver in registeredReceiver) {
            try {
                context.unregisterReceiver(receiver)
                registeredReceiver.remove(receiver)
            } catch (exception: Exception) {
                Logger.instance.error(tag, exception)
            }
        }
    }
}